File: project_euler/Problem078/Problem078.java
package project_euler;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Problem078 {
    public void solution() {
        int target = 1000000;
        List<Decimal> ways = new ArrayList<>();
        ways.add(setValue(1, 1));
        ways.add(setValue(1, 1));

        int length = 0;
        while (target >= (int) 1e5) {
            ++length;
            target /= (int) 1e5;
        }

        Decimal auxiliary = new Decimal(100, 0);
        int result = 2;
        while (true) {
            Decimal way = getWay(result, ways, auxiliary);
            if (isDivideExactly(way, length, target)) {
                System.out.println(result);
                return;
            }
            ++result;
        }
    }

    private Decimal setValue(int value, int length) {
        Decimal decimal = new Decimal(length, 0);
        decimal.value[0] = value;
        decimal.length = 1;
        return decimal;
    }

    private Decimal getWay(int target, List<Decimal> ways, Decimal auxiliary) {
        int factor = 1;
        for (int i = 1; i <= target; ++i) {
            if (!addDecimal(target - (i * (i * 3 - 1) >> 1), ways, factor, auxiliary) || !addDecimal(target - (i * (i * 3 + 1) >> 1), ways, factor, auxiliary)) {
                break;
            }
            factor *= -1;
        }
        Decimal result = new Decimal(auxiliary.length, 0);
        swapDecimal(auxiliary, result);
        ways.add(result);
        return result;
    }

    private boolean isDivideExactly(Decimal way, int length, int target) {
        if (way.length <= length) {
            return false;
        }

        for (int i = 0; i < length; ++i) {
            if (0 != way.value[i]) {
                return false;
            }
        }
        return 0 == way.value[length] % target;
    }

    private boolean addDecimal(int index, List<Decimal> ways, int factor, Decimal summation) {
        if (index < 0) {
            return false;
        }
        getSummation(ways.get(index), factor, summation);
        return true;
    }

    private void swapDecimal(Decimal previous, Decimal current) {
        int length = Math.max(previous.length, current.length);
        for (int i = 0; i < length; ++i) {
            previous.value[i] ^= current.value[i];
            current.value[i] ^= previous.value[i];
            previous.value[i] ^= current.value[i];
        }
        previous.length ^= current.length;
        current.length ^= previous.length;
        previous.length ^= current.length;
    }

    private void getSummation(Decimal decimal, int factor, Decimal summation) {
        for (int i = 0; i < decimal.length; ++i) {
            summation.value[i] += decimal.value[i] * factor;
        }

        if (factor > 0) {
            for (int i = 0; i < summation.length; ++i) {
                if (summation.value[i] >= (int) 1e5) {
                    summation.value[i + 1] += summation.value[i] / (int) 1e5;
                    summation.value[i] %= (int) 1e5;
                }
            }
            if (0 != summation.value[summation.length]) {
                ++summation.length;
            }
        } else {
            for (int i = 0; i < summation.length; ++i) {
                if (summation.value[i] < 0) {
                    --summation.value[i + 1];
                    summation.value[i] += (int) 1e5;
                }
            }
            while (0 == summation.value[summation.length - 1]) {
                --summation.length;
            }
        }
    }

    private static class Decimal {
        public int[] value;
        public int length;

        public Decimal(int capacity, int length) {
            this.value = new int[capacity];
            Arrays.fill(this.value, 0);
            this.length = length;
        }
    }
}

File: project_euler/Problem078/Problem078.go
package project_euler

import (
    "fmt"
    "math"
)

func Problem078() {
    target := 1000000
    ways := make([]Decimal, 0)
    ways = append(ways, setValue(1, 1))
    ways = append(ways, setValue(1, 1))

    length := 0
    for target >= int(1e5) {
        length++
        target /= int(1e5)
    }

    auxiliary := Decimal{make([]int, 100), 0}
    for i := 1; i < 100; i++ {
        auxiliary.value[i] = 0
    }
    result := 2
    for true {
        way := getWay(result, &ways, &auxiliary)
        if isDivideExactly(way, length, target) {
            fmt.Println(result)
            return
        }
        result++
    }
}

func setValue(value int, length int) Decimal {
    decimal := Decimal{make([]int, length), 0}
    for i := 1; i < length; i++ {
        decimal.value[i] = 0
    }
    decimal.value[0] = value
    decimal.length = 1
    return decimal
}

func getWay(target int, ways *[]Decimal, auxiliary *Decimal) Decimal {
    delta := int(math.Sqrt(float64(target * 24 + 1)))
    borders := []int{(delta + 1) / 6 + 1, (delta - 1) / 6 + 1}
    factor := 1
    for i := 1; i < borders[1]; i++ {
        getSummation(&(*ways)[target - (i * (i * 3 - 1) >> 1)], factor, auxiliary)
        getSummation(&(*ways)[target - (i * (i * 3 + 1) >> 1)], factor, auxiliary)
        factor *= -1
    }
    if borders[0] != borders[1] {
        getSummation(&(*ways)[target - (borders[1] * (borders[1] * 3 - 1) >> 1)], factor, auxiliary)
    }
    result := Decimal{make([]int, auxiliary.length), 0}
    for i := 1; i < auxiliary.length; i++ {
        result.value[i] = 0
    }
    swapDecimal(auxiliary, &result)
    *ways = append(*ways, result)
    return result
}

func isDivideExactly(way Decimal, length int, target int) bool {
    if way.length <= length {
        return false
    }

    for i := 0; i < length; i++ {
        if 0 != way.value[i] {
            return false
        }
    }
    return 0 == way.value[length] % target
}

func getSummation(decimal *Decimal, factor int, summation *Decimal) {
    for i := 0; i < decimal.length; i++ {
        summation.value[i] += decimal.value[i] * factor
    }

    if factor > 0 {
        for i := 0; i < summation.length; i++ {
            if summation.value[i] >= int(1e5) {
                summation.value[i+1] += summation.value[i] / int(1e5)
                summation.value[i] %= int(1e5)
            }
        }
        if 0 != summation.value[summation.length] {
            summation.length++
        }
    } else {
        for i := 0; i < summation.length; i++ {
            if summation.value[i] < 0 {
                summation.value[i + 1]--
                summation.value[i] += int(1e5)
            }
        }
        for 0 == summation.value[summation.length - 1] {
            summation.length--
        }
    }
}

func swapDecimal(previous *Decimal, current *Decimal) {
    length := int(math.Max(float64(previous.length), float64(current.length)))
    for i := 0; i < length; i++ {
        previous.value[i], current.value[i] = current.value[i], previous.value[i]
    }
    previous.length, current.length = current.length, previous.length
}

type Decimal struct {
    value []int
    length int
}
